rclcpp/include/rclcpp/callback_group.hpp

```diff
@@ -171,6 +174,17 @@ class CallbackGroup
   bool
   automatically_add_to_executor_with_node() const;
 
+  /// Create and return the notify guard condition.
+  RCLCPP_PUBLIC
+  std::shared_ptr<rclcpp::GuardCondition>
+  create_notify_guard_condition(
+    const rclcpp::Context::SharedPtr context_ptr);
+
+  /// Return the notify guard condition created before.
+  RCLCPP_PUBLIC
+  std::weak_ptr<rclcpp::GuardCondition>
+  get_notify_guard_condition();
```

Comment:
This is weird to me because the guard condition is stored as a shared_ptr, so why return a weak_ptr?

Reply:
Because this get_notify_guard_condition could return nullptr as the creation function is a deferred function. (I don't know what kind of this case could be.)

- use shared_ptr

```cpp
auto gc = get_notify_guard_condition();
if (gc)
   gc->trigger();
```

- use weak_ptr

```cpp
if (auto gc = get_notify_guard_condition().lock() )
   gc->trigger();
```

so I decided to use the weak_ptr for this function.